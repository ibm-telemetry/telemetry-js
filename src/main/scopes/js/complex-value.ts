/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
