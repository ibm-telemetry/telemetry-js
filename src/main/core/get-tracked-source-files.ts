/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { readFile } from 'node:fs/promises'
import * as path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from './log/logger.js'
import { TrackedFileEnumerator } from './tracked-file-enumerator.js'
import { parseDocument } from 'htmlparser2'
import type { Document as HtmlDocument } from 'domhandler'
import { HtmlParsedFile } from '../scopes/wc/interfaces.js'

/**
 * Gets all tracked source files to consider for data collection,
 * filtered by supplied file extension array.
 *
 * @param cwd - Working directory for underlying command execution.
 * @param root - Root directory in which to search for tracked source files. This is an absolute
 * path.
 * @param logger - Logger instance to use.
 * @param fileExtensions - List of file extensions to filter files by.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(
  cwd: string,
  root: string,
  logger: Logger,
  fileExtensions: string[]
) {
  logger.traceEnter('', 'getTrackedSourceFiles', [root, fileExtensions])

  const fileEnumerator = new TrackedFileEnumerator(logger)
  const files = []

  // If a file is passed instead of a directory, avoid the `git ls-tree` call
  if (fileExtensions.includes(path.extname(root))) {
    files.push(root)
  } else {
    files.push(
      ...(await fileEnumerator.find(cwd, root, (file) =>
        fileExtensions.includes(path.extname(file))
      ))
    )
  }

  const promises = files.map(async (file) => {
    return {
      fileName: file,
      async createSourceFile(): Promise<ts.SourceFile | HtmlDocument> {
        // Parse html file with `htmlparser2
        if (file.endsWith('html') || file.endsWith('htm')) {
          const parsedFile = parseDocument((await readFile(file)).toString()) as HtmlParsedFile
          parsedFile.fileName = file
        }

        // Parse any other files
        return ts.createSourceFile(
          file,
          (await readFile(file)).toString(),
          ts.ScriptTarget.ES2021,
          /* setParentNodes */ true
        )
      }
    }
  })

  const results = await Promise.all(promises)

  logger.traceExit(
    '',
    'getTrackedSourceFiles',
    results.map((result) => result.fileName)
  )
  return results
}
