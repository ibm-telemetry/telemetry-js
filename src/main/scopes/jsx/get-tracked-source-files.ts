/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from '../../core/log/logger.js'
import { TrackedFileEnumerator } from '../../core/tracked-file-enumerator.js'

/**
 * Gets all tracked source files to consider for data collection.
 *
 * @param root - Root directory in which to search for tracked source files.
 * @param logger - Logger instance to use.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(root: string, logger: Logger) {
  const fileEnumerator = new TrackedFileEnumerator(logger)
  const allowedSuffixes = ['js', 'jsx', 'ts', 'tsx']

  const files = await fileEnumerator.find(root, (file) =>
    allowedSuffixes.includes(path.extname(file))
  )

  const program = ts.createProgram(files, {})

  return program.getSourceFiles().filter((file) => !file.isDeclarationFile)
}
