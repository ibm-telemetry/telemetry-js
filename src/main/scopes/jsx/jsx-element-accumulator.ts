/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'
import { type JsxElement } from './interfaces.js'

/**
 * Responsible for maintaining an aggregated state of imports and elements.
 */
export class JsxElementAccumulator extends JsAccumulator {
  public readonly elements: JsxElement[]
  public readonly elementImports: Map<JsxElement, JsImport>

  constructor() {
    super()
    this.elements = []
    this.elementImports = new Map()
  }
}
