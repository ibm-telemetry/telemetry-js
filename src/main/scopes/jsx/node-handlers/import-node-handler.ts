/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { type ASTNodeHandler, type JsxImport } from '../interfaces.js'
import { type JsxScopeAccumulator } from '../jsx-scope-accumulator.js'

/**
 * Holds logic to construct a JsxImport object given an ImportDeclaration node.
 *
 */
export class ImportNodeHandler implements ASTNodeHandler {
  /**
   * Processes an ImportDeclaration node data and adds it to the given accumulator.
   *
   * @param node - Node element to process.
   * @param accumulator - JsxAccumulator instance that holds the aggregated imports state.
   */
  public handle(node: ts.Node, accumulator: JsxScopeAccumulator) {
    accumulator.storeImport(this.getImportData(node))
  }

  /**
   * Constructs a JsxImport object from a given ImportDeclaration type AST node.
   *
   * @param node - Node element to process.
   * @returns Constructed JsxImport object.
   */
  private getImportData(node: ts.Node): JsxImport {
    // TODO
    return {

    }
  }
}
