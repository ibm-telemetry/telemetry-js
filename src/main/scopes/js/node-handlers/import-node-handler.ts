/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { type JsxElementAccumulator } from '../../jsx/jsx-element-accumulator.js'
import { AllImportParser } from '../import-parsers/all-import-parser.js'
import { DefaultImportParser } from '../import-parsers/default-import-parser.js'
import { NamedImportParser } from '../import-parsers/named-import-parser.js'
import { RenamedImportParser } from '../import-parsers/renamed-import-parser.js'
import { JsImport } from '../interfaces.js'
import { JsNodeHandler } from './js-node-handler.js'

/**
 * Holds logic to construct a JsxImport object given an ImportDeclaration node.
 *
 */
export class ImportNodeHandler extends JsNodeHandler<JsImport[]> {
  /**
   * Processes an ImportDeclaration node data and adds it to the given accumulator.
   *
   * @param node - Node element to process.
   * @param accumulator - JsxAccumulator instance that holds the aggregated imports state.
   */
  handle(node: ts.ImportDeclaration, accumulator: JsxElementAccumulator) {
    accumulator.imports.push(...this.getData(node))
  }

  /**
   * Constructs a JsxImport object from a given ImportDeclaration type AST node.
   *
   * @param node - Node element to process.
   * @returns Constructed JsxImport object.
   */
  getData(node: ts.ImportDeclaration): JsImport[] {
    const importParsers = [
      new AllImportParser(),
      new DefaultImportParser(),
      new NamedImportParser(),
      new RenamedImportParser()
    ]

    const results: JsImport[] = []

    const importClause = node.importClause
    // This has quotes on it which need to be removed
    const importPath = node.moduleSpecifier.getText(this.sourceFile).slice(1, -1)

    if (importClause) {
      importParsers.forEach((parser) => results.push(...parser.parse(importClause, importPath)))
    }

    return results
  }
}
