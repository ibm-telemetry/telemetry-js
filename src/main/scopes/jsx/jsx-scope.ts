/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'path'
import type * as ts from 'typescript'

import { Trace } from '../../core/log/trace.js'
import { Scope } from '../../core/scope.js'
import { EmptyScopeError } from '../../exceptions/empty-scope.error.js'
import { findInstallingPackages } from '../npm/find-installing-packages.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getPackageData } from '../npm/get-package-data.js'
import { PackageData } from '../npm/interfaces.js'
import { AllImportMatcher } from './import-matchers/all-import-matcher.js'
import { NamedImportMatcher } from './import-matchers/named-import-matcher.js'
import { RenamedImportMatcher } from './import-matchers/renamed-import-matcher.js'
import { type JsxElementImportMatcher } from './interfaces.js'
import { JsxElementAccumulator } from './jsx-element-accumulator.js'
import { jsxNodeHandlerMap } from './maps/jsx-node-handler-map.js'
import { ElementMetric } from './metrics/element-metric.js'
import { SourceFileHandler } from './node-handlers/source-file-handler.js'
import { getTrackedSourceFiles } from './utils/get-tracked-source-files.js'

/**
 * Scope class dedicated to data collection from a jsx environment.
 */
export class JsxScope extends Scope {
  public override name = 'jsx' as const
  private runSync = false

  /**
   * Entry point for the scope. All scopes run asynchronously.
   */
  @Trace()
  public override async run(): Promise<void> {
    const collectorKeys = this.config.collect[this.name]
    if (collectorKeys === undefined || Object.keys(collectorKeys).length === 0) {
      throw new EmptyScopeError(this.name)
    }

    const promises: Array<Promise<void>> = []

    Object.keys(collectorKeys).forEach((key) => {
      switch (key) {
        case 'elements':
          promises.push(this.captureAllMetrics())
          break
      }
    })

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented jsx elements found in the current working
   * directory's project.
   */
  @Trace()
  async captureAllMetrics(): Promise<void> {
    const importMatchers = [
      new AllImportMatcher(),
      new NamedImportMatcher(),
      new RenamedImportMatcher()
    ]
    const instrumentedPackage = await getPackageData(this.cwd, this.cwd, this.logger)
    const sourceFiles = await this.findRelevantSourceFiles(instrumentedPackage)

    this.logger.debug('Filtered source files: ' + sourceFiles.map((f) => f.fileName))

    const promises: Promise<void>[] = []

    for (const sourceFile of sourceFiles) {
      const resultPromise = this.captureFileMetrics(
        sourceFile,
        instrumentedPackage.name,
        importMatchers
      )

      if (this.runSync) {
        await resultPromise
      } else {
        promises.push(resultPromise)
      }
    }

    await Promise.allSettled(promises)
  }

  /**
   * Finds tracked source files and then filters them based on ones that appear in a  project which
   * depends on the in-context instrumented package/version.
   *
   * @param instrumentedPackage - Data about the instrumented package to use during filtering.
   * @returns A (possibly empty) array of source files.
   */
  async findRelevantSourceFiles(instrumentedPackage: PackageData) {
    const sourceFiles = await getTrackedSourceFiles(this.root, this.logger)
    const installingPackages = await findInstallingPackages(
      this.cwd,
      this.root,
      instrumentedPackage,
      this.logger
    )

    const filterPromises = sourceFiles.map(async (f) => {
      const prefix = await getDirectoryPrefix(path.dirname(f.fileName), this.logger)
      const prefixPackageData = await getPackageData(prefix, this.root, this.logger)

      return installingPackages.some(
        (dep) => dep.name === prefixPackageData?.name && dep.version === prefixPackageData?.version
      )
    })
    const filterData = await Promise.all(filterPromises)

    return sourceFiles.filter((_, index) => {
      return filterData[index]
    })
  }

  /**
   * Generates metrics for all discovered instrumented jsx elements found
   * in the supplied SourceFile node.
   *
   * @param sourceFile - The sourcefile node to generate metrics for.
   * @param instrumentedPackageName - Name of the instrumented package to capture metrics for.
   * @param importMatchers - Matchers instances to use for import-element matching.
   */
  async captureFileMetrics(
    sourceFile: ts.SourceFile,
    instrumentedPackageName: string,
    importMatchers: JsxElementImportMatcher[]
  ) {
    const accumulator = new JsxElementAccumulator()

    this.processFile(accumulator, sourceFile)
    this.removeIrrelevantImports(accumulator, instrumentedPackageName)
    this.resolveElementImports(accumulator, importMatchers)
    await this.resolveInvokers(accumulator, sourceFile.fileName)

    accumulator.elements.forEach((jsxElement) => {
      const jsxImport = accumulator.elementImports.get(jsxElement)
      const invoker = accumulator.elementInvokers.get(jsxElement)

      if (jsxImport === undefined) {
        return
      }

      this.capture(new ElementMetric(jsxElement, jsxImport, invoker, this.config, this.logger))
    })
  }

  /**
   * Given a source file node, finds all JSX elements/imports and stores them in an accumulator.
   *
   * @param accumulator - The accumulator in which to store elements/imports.
   * @param sourceFile - Root AST node to start Jsx explorations from.
   */
  @Trace({
    argFormatter: (arg) =>
      arg instanceof JsxElementAccumulator ? '[JsxElementAccumulator]' : arg?.fileName
  })
  processFile(accumulator: JsxElementAccumulator, sourceFile: ts.SourceFile) {
    const handler = new SourceFileHandler(accumulator, jsxNodeHandlerMap, this.logger)

    handler.handle(sourceFile, sourceFile)
  }

  removeIrrelevantImports(accumulator: JsxElementAccumulator, instrumentedPackageName: string) {
    const imports = accumulator.imports.filter((jsxImport) => {
      return jsxImport.path.startsWith(instrumentedPackageName)
    })

    accumulator.imports.splice(0, accumulator.imports.length, ...imports)
  }

  resolveElementImports(
    accumulator: JsxElementAccumulator,
    elementMatchers: JsxElementImportMatcher[]
  ) {
    accumulator.elements.forEach((jsxElement) => {
      const jsxImport = elementMatchers
        .map((elementMatcher) => elementMatcher.findMatch(jsxElement, accumulator.imports))
        .find((jsxImport) => jsxImport !== undefined)

      if (!jsxImport) {
        return
      }

      accumulator.elementImports.set(jsxElement, jsxImport)
    })
  }

  /**
   * Adds data to the accumulator for each package that invokes the jsx elements in the accumulator.
   *
   * @param accumulator - Accumulator to store results in.
   * @param sourceFilePath - Absolute path to a sourceFile.
   */
  async resolveInvokers(accumulator: JsxElementAccumulator, sourceFilePath: string) {
    const containingDir = await getDirectoryPrefix(path.dirname(sourceFilePath), this.logger)

    if (containingDir === undefined) {
      return
    }

    const containingPackageData = await getPackageData(containingDir, this.root, this.logger)

    accumulator.elements.forEach((jsxElement) => {
      accumulator.elementInvokers.set(jsxElement, containingPackageData.name)
    })
  }

  /**
   * **For testing purposes only.**
   * Makes the JsxScope collection run "synchronously" (one source file at a time). Defaults to
   * `false`.
   *
   * @param runSync - Boolean of whether or not to run synchronously.
   */
  setRunSync(runSync: boolean) {
    this.runSync = runSync
  }
}
