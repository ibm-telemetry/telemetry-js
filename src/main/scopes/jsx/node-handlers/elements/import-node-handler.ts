/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { AllImportParser } from '../../import-parsers/all-import-parser.js'
import { DefaultImportParser } from '../../import-parsers/default-import-parser.js'
import { NamedImportParser } from '../../import-parsers/named-import-parser.js'
import { RenamedImportParser } from '../../import-parsers/renamed-import-parser.js'
import { type JsxImport } from '../../interfaces.js'
import { type JsxElementAccumulator } from '../../jsx-element-accumulator.js'
import { ElementNodeHandler } from './element-node-handler.js'

/**
 * Holds logic to construct a JsxImport object given an ImportDeclaration node.
 *
 */
export class ImportNodeHandler extends ElementNodeHandler<JsxImport[]> {
  /**
   * Processes an ImportDeclaration node data and adds it to the given accumulator.
   *
   * @param node - Node element to process.
   * @param accumulator - JsxAccumulator instance that holds the aggregated imports state.
   */
  handle(node: ts.ImportDeclaration, accumulator: JsxElementAccumulator) {
    accumulator.imports.push(...this.getData(node))
  }

  /**
   * Constructs a JsxImport object from a given ImportDeclaration type AST node.
   *
   * @param node - Node element to process.
   * @returns Constructed JsxImport object.
   */
  getData(node: ts.ImportDeclaration): JsxImport[] {
    const importParsers = [
      new AllImportParser(),
      new DefaultImportParser(),
      new NamedImportParser(),
      new RenamedImportParser()
    ]

    const results: JsxImport[] = []

    const importClause = node.importClause
    // This has quotes on it which need to be removed
    const importPath = node.moduleSpecifier.getText(this.sourceFile).slice(1, -1)

    if (importClause) {
      importParsers.forEach((parser) => results.push(...parser.parse(importClause, importPath)))
    }

    return results
  }
}
