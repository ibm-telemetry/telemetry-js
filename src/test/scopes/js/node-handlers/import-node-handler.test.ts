/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import ts = require('typescript')
import { describe, expect, it } from 'vitest'

import { getTrackedSourceFiles } from '../../../../main/core/get-tracked-source-files.js'
import { ImportNodeHandler } from '../../../../main/scopes/js/node-handlers/import-node-handler.js'
import { JsxElementAccumulator } from '../../../../main/scopes/jsx/jsx-element-accumulator.js'
import { JsxScope } from '../../../../main/scopes/jsx/jsx-scope.js'
import { findNodesByType } from '../../../__utils/find-nodes-by-type.js'
import { Fixture } from '../../../__utils/fixture.js'
import { initLogger } from '../../../__utils/init-logger.js'

describe('class: ImportNodeHandler', async () => {
  const logger = initLogger()
  const fixture = new Fixture('jsx-samples/all-import-types.tsx')
  const sourceFile = (
    await getTrackedSourceFiles(fixture.path, logger, JsxScope.fileExtensions)
  )[0] as ts.SourceFile
  const accumulator = new JsxElementAccumulator()
  const handler = new ImportNodeHandler(sourceFile, logger)
  it('correctly returns the JsxImports for a complex fixture', () => {
    const importDeclarations = findNodesByType(sourceFile, ts.SyntaxKind.ImportDeclaration)

    importDeclarations.forEach((importDeclaration) => {
      handler.handle(importDeclaration as ts.ImportDeclaration, accumulator)
    })

    expect(accumulator.imports).toMatchSnapshot()
  })
})
