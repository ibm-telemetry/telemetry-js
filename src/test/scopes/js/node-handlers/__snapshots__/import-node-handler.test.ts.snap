// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: ImportNodeHandler > correctly returns the JsImports for a complex fixture 1`] = `
[
  {
    "isAll": false,
    "isDefault": true,
    "name": "[Default]",
    "path": "button-library",
    "rename": "Button",
  },
  {
    "isAll": true,
    "isDefault": false,
    "name": "Accordion",
    "path": "accordion-library",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Slug",
    "path": "components-library",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Tile",
    "path": "components-library",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Notification",
    "path": "components-library",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "One",
    "path": "components-library",
    "rename": "Two",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Three",
    "path": "components-library",
    "rename": "Four",
  },
  {
    "isAll": false,
    "isDefault": true,
    "name": "[Default]",
    "path": "bla-library",
    "rename": "Bla",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Bla1",
    "path": "bla-library",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Bla2",
    "path": "bla-library",
    "rename": "Bla3",
  },
  {
    "isAll": false,
    "isDefault": true,
    "name": "[Default]",
    "path": "bla-library",
    "rename": "Bla4",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Bla1",
    "path": "bla-library",
  },
  {
    "isAll": false,
    "isDefault": false,
    "name": "Bla2",
    "path": "bla-library",
    "rename": "Bla3",
  },
]
`;
